lib/spack/spack/bootstrap.py

```diff
@@ -418,6 +418,14 @@ def _bootstrap_config_scopes():
 
 @contextlib.contextmanager
 def ensure_bootstrap_configuration():
+    # We may need to compile code from sources, so ensure we have compilers
+    # for the current platform before switching parts. We use "spack compiler
+    # list" as it searches for compilers lazily i.e. only if we don't have
+    # them already in compilers.yaml
+    compiler_cmd = spack.main.SpackCommand('compiler')
+    compiler_cmd(
+        'list', output=os.devnull, error=os.devnull, fail_on_error=False
+    )
```

Comment:
This doesn't actually ensure we have any compilers available for the current architecture -- if you have only compilers for a different OS, `spack compiler list` still won't find new ones.

We should explicitly call into the `spack.compilers` module and check whether we have any compilers for the current architecture.